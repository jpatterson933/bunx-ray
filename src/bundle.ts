// Core functions for bunx-ray MVP
import chalk from 'chalk';

export type Mod = {
  path: string;
  size: number; // bytes
};

export interface Cell {
  x: number;
  y: number;
  w: number;
  h: number;
  mod: Mod;
}

// simple slice-and-dice treemap mapped to character grid
export function treemap(mods: Mod[], W = 80, H = 24): Cell[] {
  const cells: Cell[] = [];
  let x = 0,
    y = 0,
    rowH = H;
  const total = mods.reduce((a, m) => a + m.size, 0);
  for (const m of mods) {
    const frac = m.size / total;
    const w = Math.max(1, Math.round((frac * W * H) / rowH));
    if (x + w > W) {
      y += rowH;
      x = 0;
    }
    cells.push({ x, y, w, h: rowH, mod: m });
    x += w;
  }
  return cells;
}

export function draw(cells: Cell[], W = 80, H = 24): string {
  const grid: string[][] = Array.from({ length: H }, () => Array(W).fill(' '));
  const shades = ['░', '▒', '▓', '█'];
  const max = Math.max(...cells.map((c) => c.mod.size));
  for (const c of cells) {
    const shade = shades[Math.floor((c.mod.size / max) * (shades.length - 1))];
    for (let i = 0; i < c.h; i++) {
      for (let j = 0; j < c.w; j++) {
        if (c.y + i < H && c.x + j < W) {
          grid[c.y + i][c.x + j] = shade;
        }
      }
    }
  }
  return grid.map((r) => r.join('')).join('\n');
}

// Normalize Webpack stats → Mod[]
export function normalizeWebpack(stats: any): Mod[] {
  const mods: Mod[] = [];
  if (Array.isArray(stats.modules)) {
    for (const m of stats.modules) {
      const size = m.size ?? m.parsedSize ?? 0;
      const name = m.name ?? m.identifier ?? '';
      if (size && name) mods.push({ path: name, size });
    }
  } else if (Array.isArray(stats.children)) {
    for (const child of stats.children) mods.push(...normalizeWebpack(child));
  }
  mods.sort((a, b) => b.size - a.size);
  return mods;
}

// Vite / Rollup stats (array of outputs with modules)
export function normalizeVite(stats: any): Mod[] {
  const mods: Mod[] = [];
  const outputs = Array.isArray(stats.output) ? stats.output : [stats.output ?? stats];
  for (const out of outputs) {
    if (!out || !out.modules) continue;
    const modulesObj = out.modules;
    for (const [p, m] of Object.entries<any>(modulesObj)) {
      const size = (m as any).renderedLength ?? (m as any).renderedSize ?? (m as any).originalLength ?? (m as any).size ?? 0;
      mods.push({ path: p, size });
    }
  }
  mods.sort((a, b) => b.size - a.size);
  return mods;
}

// esbuild metafile JSON
export function normalizeEsbuild(meta: any): Mod[] {
  const mods: Mod[] = [];
  if (meta.inputs && typeof meta.inputs === 'object') {
    for (const [p, info] of Object.entries<any>(meta.inputs)) {
      const size = info.bytes ?? 0;
      mods.push({ path: p, size });
    }
  }
  mods.sort((a, b) => b.size - a.size);
  return mods;
}

// ---------------- Utilities ------------------

export const SIZE_THRESHOLDS = [10 * 1024, 50 * 1024, 100 * 1024]; // bytes

export function formatSize(bytes: number): string {
  if (bytes >= 1024 * 1024) return (bytes / (1024 * 1024)).toFixed(1) + ' MB';
  if (bytes >= 1024) return (bytes / 1024).toFixed(1) + ' KB';
  return bytes + ' B';
}

export function totalSize(mods: Mod[]): number {
  return mods.reduce((a, m) => a + m.size, 0);
}

export function topModules(mods: Mod[], n = 10): Mod[] {
  return [...mods].sort((a, b) => b.size - a.size).slice(0, n);
} 