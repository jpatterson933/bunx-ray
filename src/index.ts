export type { Mod, Cell } from './bundle.js';
export {
  normalizeWebpack,
  normalizeVite,
  normalizeEsbuild,
  treemap,
  draw,
  formatSize,
  totalSize,
  topModules,
} from './bundle.js'; 