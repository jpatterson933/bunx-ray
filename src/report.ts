import { Mod, treemap, draw, formatSize, totalSize, topModules } from './bundle.js';

export interface ReportOptions {
  cols: number;
  rows: number;
  top: number;
  legend: boolean;
  summary: boolean;
  color: boolean; // reserved – color disabled for now
}

const SHADES = ['░', '▒', '▓', '█'] as const;

function calcThresholds(max: number): number[] {
  return [0.25, 0.5, 0.75].map((p) => max * p);
}

function shadeForSize(size: number, max: number): string {
  const idx = Math.floor((size / max) * (SHADES.length - 1));
  return SHADES[idx];
}

export interface RenderedReport {
  legendLine?: string;
  summaryLine?: string;
  grid: string;
  tableLines: string[];
}

export function renderReport(mods: Mod[], opts: ReportOptions): RenderedReport {
  const { cols, rows, top, legend, summary } = opts;
  const max = Math.max(...mods.map((m) => m.size));
  const thresholds = calcThresholds(max);

  // grid
  const grid = draw(treemap(mods, cols, rows), cols, rows);

  // legend
  let legendLine: string | undefined;
  if (legend) {
    legendLine =
      'Legend  ' +
      [
        `${SHADES[3]} >${formatSize(thresholds[2])}`,
        `${SHADES[2]} ${formatSize(thresholds[1])}-${formatSize(thresholds[2])}`,
        `${SHADES[1]} ${formatSize(thresholds[0])}-${formatSize(thresholds[1])}`,
        `${SHADES[0]} <${formatSize(thresholds[0])}`,
      ].join('  ');
  }

  // summary
  let summaryLine: string | undefined;
  if (summary) {
    summaryLine = `Total bundle: ${formatSize(totalSize(mods))} | modules: ${mods.length}`;
  }

  // table
  const list = topModules(mods, top);
  const tableLines: string[] = [`Top ${list.length} modules`];
  list.forEach((m, idx) => {
    const shade = shadeForSize(m.size, max);
    const pct = ((m.size / totalSize(mods)) * 100).toFixed(1).padStart(4);
    const name = m.path.length > 28 ? '…' + m.path.slice(-27) : m.path.padEnd(28);
    tableLines.push(`${(idx + 1).toString().padStart(2)} ${shade} ${name} ${formatSize(m.size).padStart(8)} (${pct}%)`);
  });

  return { legendLine, summaryLine, grid, tableLines };
} 