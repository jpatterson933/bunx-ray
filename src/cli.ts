// ---- CLI -----------------------------------------------------------------
import { Command } from 'commander';
import chalk from 'chalk';
import fs from 'fs';
import path from 'path';

import {
  Mod,
  normalizeWebpack,
  normalizeVite,
  normalizeEsbuild,
  treemap,
  draw
} from './bundle.js';

import { renderReport } from './report.js';

function main() {
  const program = new Command();

  program
    .name('bunx-ray')
    .description('ASCII heat-map bundle viewer')
    .argument('<stats>', 'Build stats JSON file')
    .option('--webpack', 'Input is Webpack stats (default)')
    .option('--vite', 'Input is Vite/Rollup stats')
    .option('--esbuild', 'Input is esbuild metafile')
    .option('--cols <number>', 'Terminal columns (default 80)', '80')
    .option('--rows <number>', 'Terminal rows (default 24)', '24')
    .option('--top <number>', 'Show N largest modules (default 10)', '10')
    .option('--no-legend', 'Hide legend line')
    .option('--no-summary', 'Hide summary line')
    .option('--grid-only', 'Only print grid (implies --no-legend --no-summary)')
    .option('--demo', 'Render built-in demo heat-map')
    .parse(process.argv);

  const opts = program.opts();
  const file = program.args[0];

  const cols = Number(opts.cols);
  const rows = Number(opts.rows);
  if (!Number.isFinite(cols) || !Number.isFinite(rows)) {
    console.error(chalk.red('Error: --cols and --rows must be numbers'));
    process.exit(1);
  }

  const raw = fs.readFileSync(path.resolve(process.cwd(), file), 'utf8');
  let stats: any;
  try {
    stats = JSON.parse(raw);
  } catch (e) {
    console.error(chalk.red(`Failed to parse JSON from ${file}`));
    process.exit(1);
  }

  // Choose adapter based on flags or auto-detect.
  let mods: Mod[] = [];
  if (opts.webpack) {
    mods = normalizeWebpack(stats);
  } else if (opts.vite) {
    mods = normalizeVite(stats);
  } else if (opts.esbuild) {
    mods = normalizeEsbuild(stats);
  } else {
    // auto-detect simple heuristics
    if (stats.inputs && stats.outputs) mods = normalizeEsbuild(stats);
    else if (stats.modules || stats.children) mods = normalizeWebpack(stats);
    else if (stats.output) mods = normalizeVite(stats);
    else {
      console.error(chalk.red('Unable to detect stats format; please pass --webpack | --vite | --esbuild'));
      process.exit(1);
    }
  }

  if (mods.length === 0) {
    console.error(chalk.yellow('No modules found in stats file.'));
    process.exit(1);
  }

  const report = renderReport(mods, {
    cols,
    rows,
    top: Number(opts.top ?? 10),
    legend: opts.legend !== false && !opts.gridOnly,
    summary: opts.summary !== false && !opts.gridOnly,
    color: true,
  });

  if (report.legendLine) console.log(report.legendLine);
  if (report.summaryLine) console.log(report.summaryLine);

  console.log('\n' + report.grid + '\n');

  report.tableLines.forEach((l) => console.log(l));
}

main(); 