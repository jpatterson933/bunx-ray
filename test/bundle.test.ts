/// <reference types="vitest" />

import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { normalizeWebpack, normalizeVite, normalizeEsbuild, treemap, draw } from '../src/bundle';

const fixturesDir = new URL('../fixtures/', import.meta.url).pathname;

function load(name: string) {
  return JSON.parse(readFileSync(`${fixturesDir}${name}`, 'utf8'));
}

describe('normalize + draw', () => {
  it('webpack fixture renders', () => {
    const mods = normalizeWebpack(load('webpack-sample.json'));
    const ascii = draw(treemap(mods, 40, 8), 40, 8);
    expect(ascii).toMatchSnapshot();
  });

  it('vite fixture renders', () => {
    const mods = normalizeVite(load('vite-sample.json'));
    const ascii = draw(treemap(mods, 40, 8), 40, 8);
    expect(ascii).toMatchSnapshot();
  });

  it('esbuild fixture renders', () => {
    const mods = normalizeEsbuild(load('esbuild-sample.json'));
    const ascii = draw(treemap(mods, 40, 8), 40, 8);
    expect(ascii).toMatchSnapshot();
  });
}); 