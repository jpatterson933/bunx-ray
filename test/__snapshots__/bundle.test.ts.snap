// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`normalize + draw > esbuild fixture renders 1`] = `
"███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒"
`;

exports[`normalize + draw > vite fixture renders 1`] = `
"███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒
███████████████████▓▓▓▓▓▓▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒"
`;

exports[`normalize + draw > webpack fixture renders 1`] = `
"█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░
█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░
█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░
█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░
█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░
█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░
█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░
█████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░"
`;
